### ISSUE TITLE
bbox mAP =0. I used the same dataset on the yolact, and the result is 60+. 

### ISSUE THREAD

**Issue by jonguo111** [open-mmlab/mmdetection#2361]
![image](https://user-images.githubusercontent.com/43916561/77986928-5f970a00-72e6-11ea-8ee2-c8e1bd3e81c1.png)

![image](https://user-images.githubusercontent.com/43916561/77986902-48f0b300-72e6-11ea-8efa-33fbbc695ab8.png)
could you pls help me figure it out? the dataset is coco format and there are 4 classes.
Thank you in advance.

**Comment by DoveJay** [comment]
I face the same problem, please give me a feedback

**Comment by jonguo111** [comment]
> I face the same problem, please give me a feedback

i have no idea....

**Comment by DoveJay** [comment]
> > I face the same problem, please give me a feedback
> 
> i have no idea....

a ou

**Comment by jonguo111** [comment]
> > > I face the same problem, please give me a feedback
> > 
> > 
> > i have no idea....
> 
> a ou

I can visualize the detection results, but the mAP values are zero...... So weird

**Comment by DoveJay** [comment]
this is my social media number 962159331@qq.com
we can communicate with each other

**Comment by hellock** [comment]
Please follow the `Error report` issue template.

**Comment by jonguo111** [comment]
**Environment**
sys.platform: linux
Python: 3.7.7 (default, Mar 26 2020, 15:48:22) [GCC 7.3.0]
CUDA available: True
CUDA_HOME: /usr/local/cuda-9.0
NVCC: Cuda compilation tools, release 9.0, V9.0.176
GPU 0: Tesla K80
GCC: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
PyTorch: 1.1.0
PyTorch compiling details: PyTorch built with:
  - GCC 4.9
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
  - OpenMP 201307 (a.k.a. OpenMP 4.0)
  - NNPACK is enabled
  - CUDA Runtime 9.0
  - NVCC architecture flags: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_50,code=compute_50
  - CuDNN 7.5.1
  - Magma 2.5.0
  - Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS=  -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=True, USE_NNPACK=True, USE_OPENMP=ON,

TorchVision: 0.3.0
OpenCV: 4.2.0
MMCV: 0.4.2
MMDetection: 1.1.0+1d877f8
MMDetection Compiler: GCC 5.4
MMDetection CUDA Compiler: 9.0
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
**Reproduction**
1. What command or script did you run?
for training: python tools/train.py configs/mask_rcnn_r101_fpn_1x.py
for testing: python tools/test.py configs/mask_rcnn_r101_fpn_1x.py  work_dirs/mask_rcnn_r101_fpn_1x/epoch_1.pth --eval bbox

2. Did you make any modifications on the code or config? Did you understand what you have modified?
Yes. I modified the **coco.py** and the modification was **the change of i+1 to i**. It can be seen in the following pic. Also, I changed to my custom data numbers in **class_names.py** and **coco.py**
![image](https://user-images.githubusercontent.com/43916561/78165566-8144de00-7419-11ea-9c18-8919741607c2.png)

3. What dataset did you use?
My custom dataset was labeled with the labelme

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
**Describe the bug**
I can visualize the detection results. They were good. However, the mAP values are zero. I tested on the yolact. They are over 60( ALL). Pls see the pic in the following part. Even the last epoch, it still shows zero
![image](https://user-images.githubusercontent.com/43916561/78166536-f369f280-741a-11ea-9e17-58234003f425.png)

**Comment by jonguo111** [comment]
> Please follow the `Error report` issue template.
Environment
sys.platform: linux
Python: 3.7.7 (default, Mar 26 2020, 15:48:22) [GCC 7.3.0]
CUDA available: True
CUDA_HOME: /usr/local/cuda-9.0
NVCC: Cuda compilation tools, release 9.0, V9.0.176
GPU 0: Tesla K80
GCC: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
PyTorch: 1.1.0
PyTorch compiling details: PyTorch built with:

GCC 4.9
Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
Intel(R) MKL-DNN v0.18.1 (Git Hash 7de7e5d02bf687f971e7668963649728356e0c20)
OpenMP 201307 (a.k.a. OpenMP 4.0)
NNPACK is enabled
CUDA Runtime 9.0
NVCC architecture flags: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_50,code=compute_50
CuDNN 7.5.1
Magma 2.5.0
Build settings: BLAS=MKL, BUILD_TYPE=Release, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=True, USE_NNPACK=True, USE_OPENMP=ON,
TorchVision: 0.3.0
OpenCV: 4.2.0
MMCV: 0.4.2
MMDetection: 1.1.0+1d877f8
MMDetection Compiler: GCC 5.4
MMDetection CUDA Compiler: 9.0
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Reproduction

What command or script did you run?
for training: python tools/train.py configs/mask_rcnn_r101_fpn_1x.py
for testing: python tools/test.py configs/mask_rcnn_r101_fpn_1x.py work_dirs/mask_rcnn_r101_fpn_1x/epoch_1.pth --eval bbox

Did you make any modifications on the code or config? Did you understand what you have modified?
Yes. I modified the coco.py and the modification was the change of i+1 to i. It can be seen in the following pic. Also, I changed to my custom data numbers in class_names.py and coco.py

![image](https://user-images.githubusercontent.com/43916561/78166792-5a87a700-741b-11ea-9a65-a7d23ba61fce.png)

What dataset did you use?
My custom dataset was labeled with the labelme

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Describe the bug
I can visualize the detection results. They were good. However, the mAP values are zero. I tested on the yolact. They are over 60( ALL). Pls see the pic in the following part. Even the last epoch, it still shows zero

![image](https://user-images.githubusercontent.com/43916561/78166842-70956780-741b-11ea-8443-79c5fda33811.png)

**Comment by jonguo111** [comment]
I solved this issue. The reason is the category_id in output json is wrong. They begin with 0 other than 1 in my case. 
Firstly, I saved the bbox json and segm json from the following command. I established a new file named results in the root folder. After you run the python tools/test.py configs/xxxx/ work_dirs/xxxxx/, the bbox json and segm json will be saved in the new folder. Then you need to correct your json files(changing the category_id). In the end, you can evaluate your mAP in the newly established eval.py in the following part. 

+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
![image](https://user-images.githubusercontent.com/43916561/78209227-6f445900-7474-11ea-93b5-054e011debdf.png)
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
# -*- coding: utf-8 -*-
"""
Created on Wed Apr  1 16:48:07 2020

@author: Feng
"""

import argparse

from pycocotools.coco import COCO
from pycocotools.cocoeval import COCOeval


parser = argparse.ArgumentParser(description='COCO Detections Evaluator')
parser.add_argument('--bbox_det_file', default='/home/fengg/mmdetection/results/bbox_final.json', type=str)
parser.add_argument('--mask_det_file', default='/home/fengg/mmdetection/results/segm_final.json', type=str)
parser.add_argument('--gt_ann_file',   default='/home/fengg/mmdetection/data/annotations/instances_val2017.json', type=str)
parser.add_argument('--eval_type',     default='both', choices=['bbox', 'mask', 'both'], type=str)
args = parser.parse_args()



if __name__ == '__main__':

	eval_bbox = (args.eval_type in ('bbox', 'both'))
	eval_mask = (args.eval_type in ('mask', 'both'))

	print('Loading annotations...')
	gt_annotations = COCO(args.gt_ann_file)
	if eval_bbox:
		bbox_dets = gt_annotations.loadRes(args.bbox_det_file)
	if eval_mask:
		mask_dets = gt_annotations.loadRes(args.mask_det_file)

	if eval_bbox:
		print('\nEvaluating BBoxes:')
		bbox_eval = COCOeval(gt_annotations, bbox_dets, 'bbox')
		bbox_eval.evaluate()
		bbox_eval.accumulate()
		bbox_eval.summarize()
	
	if eval_mask:
		print('\nEvaluating Masks:')
		bbox_eval = COCOeval(gt_annotations, mask_dets, 'segm')
		bbox_eval.evaluate()
		bbox_eval.accumulate()
		bbox_eval.summarize()

**Comment by ZwwWayne** [comment]
Hi @jonguo111 ，
The problem might lie in that you change the labels' id to `i` rather than `i+1`.
![image](https://user-images.githubusercontent.com/40779233/78317492-fdaffc00-7594-11ea-9237-5f54fa4a3b8c.png)
Current MMDetection system takes the convention that the background has category id 0, and the foreground's category id ranges from `[1, num_classes]`. Therefore, change labels' id to `i` from `i+1`
will cause the inconsistency between your dataloader and the models target assignment. 
The change is not that simple, you can see this PR https://github.com/open-mmlab/mmdetection/pull/2374. We plan to do that in the V2.0 version and it almost changes the whole codebase.

**Comment by xanjay** [comment]
In my case, category id in train json and val json were different by mistake. I could visualize the detection results, but the mAP values were zeros.
